File: foothold_app/android/app/src/main/java/dev/rohit/foothold/capture/WaNotificationListener.java
package dev.rohit.foothold.capture;

import android.app.Notification;
import android.os.Bundle;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;

/**
 * Reads incoming WhatsApp notifications; for a TAGGED lead (name resolved via
 * the synced registry, incl. Hindi names) it POSTs the message to
 * /api/v1/capture/inbound, which scores it. Inbound-only; requires the user to
 * grant "Notification access".
 */
public class WaNotificationListener extends NotificationListenerService {
    private CaptureConfig cfg;
    private CaptureApi api;

    @Override
    public void onCreate() {
        super.onCreate();
        cfg = new CaptureConfig(this);
        api = new CaptureApi(cfg);
    }

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        String pkg = sbn.getPackageName();
        if (!"com.whatsapp".equals(pkg) && !"com.whatsapp.w4b".equals(pkg)) return;

        Bundle ex = sbn.getNotification().extras;
        if (ex == null) return;
        CharSequence titleCs = ex.getCharSequence(Notification.EXTRA_TITLE);
        CharSequence textCs = ex.getCharSequence(Notification.EXTRA_TEXT);
        if (titleCs == null || textCs == null) return;

        String title = titleCs.toString();
        String text = textCs.toString();
        if (text.startsWith("📞")) return;          // skip call notifications (📞)

        String key = title + "|" + text;
        if (key.equals(cfg.lastKey())) return;                 // dedupe repeats
        cfg.setLastKey(key);

        String digits = cfg.digitsForName(title);
        if (digits == null) return;                            // PRIVACY: tagged leads only
        api.capture(digits, "whatsapp", text.length() > 200 ? text.substring(0, 200) : text, "in");
    }
}

File: foothold_app/android/capture/WaNotificationListener.kt
package dev.rohit.foothold.capture

import android.app.Notification
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification

/**
 * Reads incoming WhatsApp notifications and turns messages from TAGGED leads into
 * captures POSTed to /api/v1/capture/inbound (channel=whatsapp, direction=in). The
 * heavier intent extraction happens server-side; here we just forward the text.
 *
 * Honest limits (unchanged from the design in project_realestate_foothold):
 *  - Inbound only (notifications don't carry the salesman's outbound).
 *  - WhatsApp notifications carry the contact DISPLAY NAME, not a number, so we
 *    resolve name → tagged number via the synced registry. Full history needs the
 *    WhatsApp Business API.
 *  - Requires "Notification access" (special settings toggle).
 *  - Tagged leads only — every other chat is ignored on-device.
 *
 * ⚠ WRITTEN, NOT DEVICE-VERIFIED here.
 */
class WaNotificationListener : NotificationListenerService() {
    private val cfg by lazy { CaptureConfig(this) }
    private val api by lazy { CaptureApi(cfg) }
    private var lastKey: String = ""

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val pkg = sbn.packageName
        if (pkg != "com.whatsapp" && pkg != "com.whatsapp.w4b") return

        val ex = sbn.notification.extras
        val title = ex.getCharSequence(Notification.EXTRA_TITLE)?.toString()   // sender / group name
        val text = ex.getCharSequence(Notification.EXTRA_TEXT)?.toString() ?: return

        if (title.isNullOrBlank() || text.startsWith("📞")) return             // skip calls/typing
        val key = "$title|$text"
        if (key == lastKey) return; lastKey = key

        val digits = cfg.digitsForName(title) ?: return   // PRIVACY: only tagged leads
        api.capture(digits, "whatsapp", text.take(200), "in")
    }
}
